Eat together 서비스 구현

jjj0611 opened the pull request:
[x] Slack 요청을 위한 SlackApiService 분리
[x] Eat together 기능을 위한 EatTogetherService 구현
[x] User 한명의 정보를 가지고 있는 UserProfile 생성
[x] EatTogether 정보를 가지고 있는 UserMatcher 생성

To-do// Scheduling 기능 추가
[ ] 11시 30분, 18시 전에 UserMatcher.generateMatchedUser 호출
[ ] generateMatchedUser 호출이 성공하면 resetAppliedUser 호출

moonyoungCHAE committed:
feat. 같이 먹어요 기능 추가

- 점심, 저녁 시간 범위 확인
- 이미 등록한 사용자인지 확인

moonyoungCHAE committed:
feat: 같이 먹어요 매칭 기능 추가

- 특정 시간에 사용자 매칭해주는 기능 추가해야 된다.

moonyoungCHAE committed:
feat: 오늘 뭐먹지?를 같이 먹어요, 결과 확인과 같은 줄로 이동

moonyoungCHAE committed:
feat: User Service 생성

- id to display name
- 모든 user list 가져오기

jjj0611 committed:
[feature] EatTogetherService 완성

	- SlackApiService 분리
	- EatTogetherService 구현

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/domain/UserMatcher.java:
```diff
@@ -0,0 +1,50 @@
+@Component
+public class UserMatcher {
+
+    public static final int LAST_INDEX = -1;
```
이거 안된다~ㅎㅋㅎㅎㅎㅋ

jjj0611 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/domain/UserMatcher.java:
```diff
@@ -0,0 +1,50 @@
+    private Set<UserProfile> appliedUsers = new HashSet<>();
+    private List<List<UserProfile>> matchedUsers = new ArrayList<>();
+
+    public boolean add(UserProfile userProfile) {
```
apply하는 user에 add하는 것인지, match된 user에 add하는 것인지 애매한 것 같음
addAppliedUser????..

jjj0611 replied:
그냥 apply라고만 하면 어때?

moonyoungCHAE replied:
굳굳굳

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/domain/UserMatcher.java:
```diff
@@ -0,0 +1,50 @@
+        }
+        List<UserProfile> userProfiles = new ArrayList<>(appliedUsers);
+        Collections.shuffle(userProfiles);
+        List<List<UserProfile>> matchedUsers = Lists.partition(userProfiles, 4);
```
```
        LinkedList<List<UserProfile>> matchedUsers = new LinkedList<>(Lists.partition(userProfiles, 4));
        List<UserProfile> lastTeam = matchedUsers.getLast();
        if (lastTeam.size() == 1) {
            matchedUsers.removeLast();
            matchedUsers.getLast().addAll(lastTeam);
        }
```
어떰??!

jjj0611 replied:
좋아요 👍

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/service/EatTogetherService.java:
```diff
@@ -0,0 +1,86 @@
+    private final UserMatcher userMatcher;
+
+    private static Map<String, List<LocalTime>> initializeTimeScope() {
+        Map<String, List<LocalTime>> timeScope = new HashMap<>();
```
시간과 관련된 부분이 지저분한 것 같음 ! 악
찾아봤는데 시간의 범위와 관련된 게
```Duration```은 몇 시, 몇 분 몇 초만큼 차이나는지
```Period```는 날짜까지만 제공해서 쓸 수 있는 건 딱히 없는 듯

따로 클래스를 만들까? 그냥 둘까?

jjj0611 replied:
시간 관련된 로직이 많으면 분리하는게 좋긴 할듯?

너무 Service가 많아진다 생각하면 현재 상태를 유지해도 되고,
Service 많아져도 명확하게 역할을 구분하고 싶다고 하면 추가해도 되고!

어떻게 나을거 같아?

moonyoungCHAE replied:
난 따로가 좋은 것 같음!
근데 시간 범위와 관련된 api가 없다는 게 납득이 안됔ㅋㅋㅋㅋ 좀만 더 찾아볼겡

moonyoungCHAE replied:
```
    private static final TemporalQuery<Boolean> WorkingHoursQuery =
              temporal -> {
                  LocalTime t = LocalTime.from(temporal);
                  return t.compareTo(LocalTime.of(9, 0)) >= 0 &&
                            t.compareTo(LocalTime.of(17, 0)) < 0;
              };
```
이 방식으로 수정함 !!

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/slack/dto/response/block/BlockResponseFactory.java:
```diff
@@ -48,10 +48,11 @@ public static Block createInitActionBlock(String blockId) {
     }
 
-    public static Block createTogetherActionBlock(String blockId) {
+    public static Block createRecoomendAndEatTogetherActionBlock(String blockId) {
```
```RecoomendAndEatTogether``` 보다는 어제 말한 ```SpeicalService```가 더 좋은 것 같음..
다른 거 할 만한 거 뭐 있을까....

jjj0611 replied:
SpecialService 괜찮다 👍

moonyoungCHAE commented on src/main/java/org/woowacourse/lunchbot/service/SlackApiService.java:
```diff
@@ -0,0 +1,85 @@
+
+@Slf4j
+@Service
+public class SlackApiService {
```
Service도 Singleton으로 관리가 되니까
```
@Component
@Scope("request")
```
이게 더 어울리는 것 같은데 맞아? 확인 부탁!
request : HTTP 요청별로 인스턴스화 되며 요청이 끝나면 소멸 (spring mvc webapplication 용도)

jjj0611 replied:
그럼 다른 곳에서 slackApiService.() 메서드 호출할 때마다 인스턴스를 만드는건가?
어떤 이점이 있을까?

moonyoungCHAE replied:
하나의 인스턴스로 여러 명의 사용자를 처리하면,
웹클라이언트가 통신중에 다른 사용자가 접근하면 문제가 생길 것 같아서!!

jjj0611 replied:
이건 아마 스프링에서 Queue 같은거 사용해서 자동으로 분배 해줄걸??

moonyoungCHAE replied:
아하 그렇군 인텔리jjjjjjjjj
```
@Component
@Scope("request")
```
이건 스프링에서 사용자쪽으로 request 보낼 때 쓰나봐

moonyoungCHAE reviewed (commented):
1. 식당 정보 얼마에 한 번씩 update할까?
지금 ``` Restaurants ``` - ```initializeRestaurants``` 에서
식당 정보 받아오기 + 오늘의 추천 생성 동시에 하는데,
둘다 하루에 한 번 업데이트ㄱㄱ?
아님 따로 ???

2. 식당 제보 받는 거 할 말?
구글 설문지에서 구글 스프레드 시트로 바로 데이터 받아지는 거 가능

jjj0611 committed:
[refactoring] 변수명 및 로직 변경

	- eat together 사용자 받을 때 add 메서드 명을 apply로 변경
	- LinkedList를 활용해서 마지막 노드를 쉽게 찾을 수 있도록 변경
	- recommand와 eat together 서비스를 합쳐서 special service로 변경

moonyoungCHAE reviewed (approved):
완벼쿠
